Extension/src/LanguageServer/client.ts

```diff
@@ -893,12 +978,43 @@ export class DefaultClient implements Client {
 
                     if (isFirstClient) {
                         workspaceReferences = new refs.ReferencesManager(this);
-
                         // The configurations will not be sent to the language server until the default include paths and frameworks have been set.
                         // The event handlers must be set before this happens.
-                        const inputCompilerDefaults: configs.CompilerDefaults = await languageClient.sendRequest(QueryCompilerDefaultsRequest, {});
+                        const compilerPath: PersistentState<string[]> = new PersistentState<string[]>("CPP.trustedCompilerPaths", []);
+                        compilerPath.Value = [];
+                        const inputCompilerDefaults: configs.CompilerDefaults = await this.requestCompiler(compilerPath.Value);
                         compilerDefaults = inputCompilerDefaults;
                         this.configuration.CompilerDefaults = compilerDefaults;
+                        this.configuration.setupConfigurations();
+                        if (!compilerDefaults.trustedCompilerFound) {
+                            this.notifyWhenLanguageClientReady(() => {
+                                // if there is a compilerPath in c_cpp_properties.json, make compiler trusted and query compiler
+                                clients.forEach(async client => {
```

Comment:
I'm not sure this is the right place for this.  Since the configuration logic will trust any explicitly set `compilerPath` it sees, these checks seem redundant.  It seems like we need a place where, after configurations have been processed for all clients, we can simply check if there is a trusted compiler yet, and prompt if there is not...  It looks like just be a matter of deferring this until all clients have been constructed.  Maybe the check to display the prompt could be moved into `activate` in extension.ts, right after `clients = new ClientCollection();`.

Reply:
It might also be worth setting a timer and not doing the check until after a few seconds, to give a custom configuration provider the opportunity to start up (which does not happen immediately).  When/if a custom configuration provider registeres with cpptools, it may provide a browse config with a (trustable) compiler path in it.